app/src/main/java/com/gogreen/greenmachine/main/MainActivity.java

```diff
@@ -780,28 +734,95 @@ protected void onPreExecute() {
+                for (ParseGeoPoint p : result) {
+                    origins += p.getLatitude() + "," + p.getLongitude() + "|";
                 }
+                origins = origins.length()>0 ? (origins.substring(0, origins.length() - 1)) : "";
+                return new Tuple<String,Marker[]>(origins,(params[0].y));
+
+            } catch (ParseException e) {
+                e.printStackTrace();
+                return null;
             }
-            return origins;
+
         }
 
         @Override
-        protected void onPostExecute(String result) {
+        protected void onPostExecute(Tuple<String,Marker[]> result){
             super.onPostExecute(result);
             pdLoading.dismiss();
+            if (result==null)
+                return;
+
+            Marker m=(result.y)[0];
+            String origins=result.x;
+            if (origins=="") {
+                (result.y)[0].setTitle("Next Pickup: N/A");
+                (result.y)[0].setSnippet("no drivers yet");
+                return;
+            }
+            int minTime=24*60*60;  //pickupTime in seconds
+
+            //pack the query for distance matrix
+            String destinations=Double.toString((result.y)[0].getPosition().latitude)+","+Double.toString((result.y)[0].getPosition().longitude);
+            String mode="driving";
+            String language="US-EN";
+            String key=getString(R.string.google_maps_key);
+            String urlString="https://maps.googleapis.com/maps/api/distancematrix/json";
+
+            GenericUrl url = new GenericUrl(urlString);
+            url.put("origins", origins);
+            url.put("destinations", destinations);
+            url.put("mode", mode);
+            url.put("language", language);
+            url.put("key", key);
+            Log.i(MainActivity.class.getSimpleName(), url.toString());
+            try {
+                Result r = new RetrieveDistanceMatrix().execute(url).get();
+                List<Row> rows = r.rows;
+                for (Row row : rows) {
+                    List<Element> elements = row.elements;
+                    for (Element e : elements) {
+                        minTime=Math.min(minTime,e.duration.value);
+                    }
+                }
+
+            }
+            catch (Exception e){
+                e.printStackTrace();
+                return;
+            }
+            Log.i(MainActivity.class.getSimpleName(), Integer.toString(minTime));
+            //if minTime has not been reduced, possibly no drivers on the hotspot
+            if (minTime<1440){
+                (result.y)[0].setTitle("Next Pickup: "+(minTime/60)+" min");
+                (result.y)[0].setSnippet("Text @ 650-290-2120");
+            }
+            else {
+                (result.y)[0].setTitle("Next Pickup: N/A");
+                (result.y)[0].setSnippet("no drivers yet");
+            }
+            return;
         }
     }
+}
+
+class Tuple<X, Y> {
```

Comment:
add explicit "protected" or "public". Java by default is protected if you leave it out but please be explicit.